brave-instrumentation-jaxrs throws UnsupportedOperationException on the latest Jersey Client

## Describe the Bug
I recently updated my brave-instumentation-jaxrs dependency from version 4.19.3 to 5.9.0 and noticed that with no additional code change brave instrumentation started failing with UnsupportedOperationException.

## Steps to Reproduce
Steps to reproduce the behavior:

The minimal example to reproduce this:

```
import brave.Tracing;
import brave.http.HttpTracing;
import brave.jaxrs2.TracingClientFilter;
import org.glassfish.jersey.client.ClientConfig;
import org.glassfish.jersey.client.JerseyClientBuilder;

Tracing tracing = Tracing.newBuilder().localServiceName("example").build()
ClientConfig clientConfig = new ClientConfig();
clientConfig.register(TracingClientFilter.create(HttpTracing.create(tracing).clientOf("any")));
Client client = JerseyClientBuilder.createClient(clientConfig);
```

Then try to hit an endpoint of any kind.

You'll get an exception that looks like this:
```
javax.ws.rs.ProcessingException: null
	at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:263)
	at org.glassfish.jersey.client.JerseyInvocation.lambda$invoke$1(JerseyInvocation.java:750)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:205)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:390)
	at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:748)
	at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:404)
	at org.glassfish.jersey.client.JerseyInvocation$Builder.get(JerseyInvocation.java:300)
        ...
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.mockito.internal.junit.JUnitRule$1.evaluateSafely(JUnitRule.java:57)
	at org.mockito.internal.junit.JUnitRule$1.evaluate(JUnitRule.java:48)
	at au.com.dius.pact.consumer.BaseProviderRule.lambda$runPactTest$2(BaseProviderRule.java:160)
	at au.com.dius.pact.consumer.BaseMockServer.runAndWritePact(MockHttpServer.kt:181)
	at au.com.dius.pact.consumer.ConsumerPactRunnerKt.runConsumerTest(ConsumerPactRunner.kt:13)
	at au.com.dius.pact.consumer.BaseProviderRule.runPactTest(BaseProviderRule.java:158)
	at au.com.dius.pact.consumer.BaseProviderRule.access$100(BaseProviderRule.java:22)
	at au.com.dius.pact.consumer.BaseProviderRule$1.evaluate(BaseProviderRule.java:78)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
Caused by: java.lang.UnsupportedOperationException: null
	at java.util.AbstractMap.put(AbstractMap.java:209)
	at javax.ws.rs.core.AbstractMultivaluedMap.getValues(AbstractMultivaluedMap.java:278)
	at javax.ws.rs.core.AbstractMultivaluedMap.putSingle(AbstractMultivaluedMap.java:96)
	at brave.jaxrs2.TracingClientFilter$HttpClientRequest.header(TracingClientFilter.java:103)
	at brave.http.HttpClientRequest$1.put(HttpClientRequest.java:31)
	at brave.http.HttpClientRequest$1.put(HttpClientRequest.java:29)
	at brave.propagation.B3Propagation$B3Injector.injectMulti(B3Propagation.java:202)
	at brave.propagation.B3Propagation$B3Injector.inject(B3Propagation.java:195)
	at brave.http.HttpClientHandler.handleSend(HttpClientHandler.java:113)
	at brave.http.HttpClientHandler.handleSend(HttpClientHandler.java:101)
	at brave.jaxrs2.TracingClientFilter.filter(TracingClientFilter.java:64)
	at org.glassfish.jersey.client.ClientFilteringStages$RequestFilteringStage.apply(ClientFilteringStages.java:91)
	at org.glassfish.jersey.client.ClientFilteringStages$RequestFilteringStage.apply(ClientFilteringStages.java:79)
	at org.glassfish.jersey.process.internal.Stages.process(Stages.java:147)
	at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:254)
	... 42 common frames omitted
```

## Expected Behaviour
Request successfully executed and being traced propertly.

--- a/instrumentation/jaxrs2/src/main/java/brave/jaxrs2/TracingClientFilter.java
+++ b/instrumentation/jaxrs2/src/main/java/brave/jaxrs2/TracingClientFilter.java
@@ -100,7 +100,7 @@ public final class TracingClientFilter implements ClientRequestFilter, ClientRes
     }
 
     @Override public void header(String name, String value) {
-      delegate.getStringHeaders().putSingle(name, value);
+      delegate.getHeaders().putSingle(name, value);
     }
   }
 

--- a/instrumentation/jaxrs2/src/test/java/brave/jaxrs2/HttpClientRequestTest.java
+++ b/instrumentation/jaxrs2/src/test/java/brave/jaxrs2/HttpClientRequestTest.java
@@ -1,0 +1,68 @@
+/*
+ * Copyright 2013-2019 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.jaxrs2;
+
+import brave.jaxrs2.TracingClientFilter.HttpClientRequest;
+import org.jboss.resteasy.core.Headers;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import javax.ws.rs.client.ClientRequestContext;
+import javax.ws.rs.core.MultivaluedMap;
+import java.net.URI;
+import java.util.Collections;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.entry;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class HttpClientRequestTest {
+  @Mock ClientRequestContext request;
+
+  @Test public void method() {
+    when(request.getMethod()).thenReturn("GET");
+
+    assertThat(new HttpClientRequest(request).method()).isEqualTo("GET");
+  }
+
+  @Test public void path() {
+    when(request.getUri()).thenReturn(URI.create("http://localhost/api"));
+
+    assertThat(new HttpClientRequest(request).path()).isEqualTo("/api");
+  }
+
+  @Test public void url() {
+    when(request.getUri()).thenReturn(URI.create("http://localhost/api"));
+
+    assertThat(new HttpClientRequest(request).url()).isEqualTo("http://localhost/api");
+  }
+
+  @Test public void header() {
+    when(request.getHeaderString("name")).thenReturn("value");
+
+    assertThat(new HttpClientRequest(request).header("name")).isEqualTo("value");
+  }
+
+  @Test public void putHeader() {
+    MultivaluedMap<String, Object> headers = new Headers<>();
+    when(request.getHeaders()).thenReturn(headers);
+
+    new HttpClientRequest(request).header("name", "value");
+
+    assertThat(headers).containsExactly(entry("name", Collections.singletonList("value")));
+  }
+}

--- a/instrumentation/jaxrs2/src/test/java/brave/jaxrs2/ITTracingJaxRSClientBuilder.java
+++ b/instrumentation/jaxrs2/src/test/java/brave/jaxrs2/ITTracingJaxRSClientBuilder.java
@@ -1,0 +1,98 @@
+/*
+ * Copyright 2013-2019 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.jaxrs2;
+
+import brave.Span;
+import brave.test.http.ITHttpAsyncClient;
+import org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder;
+import org.junit.Ignore;
+
+import javax.ws.rs.ProcessingException;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.client.InvocationCallback;
+import javax.ws.rs.core.MediaType;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+public class ITTracingJaxRSClientBuilder extends ITHttpAsyncClient<Client> {
+  @Override protected Client newClient(int port) {
+    return new ResteasyClientBuilder()
+      .register(TracingClientFilter.create(httpTracing))
+      .connectTimeout(1, TimeUnit.SECONDS)
+      .readTimeout(1, TimeUnit.SECONDS)
+      .executorService(httpTracing.tracing().currentTraceContext()
+        .executorService(Executors.newCachedThreadPool()))
+      .build();
+  }
+
+  @Override protected void closeClient(Client client) {
+    client.close();
+  }
+
+  @Override protected void get(Client client, String pathIncludingQuery) {
+    try {
+      client.target(url(pathIncludingQuery))
+        .request(MediaType.TEXT_PLAIN_TYPE)
+        .get(String.class);
+    } catch (ProcessingException ex) {
+      Span span = tracer().currentSpan();
+      if (null != span) {
+        span.error(ex).finish();
+      }
+      throw ex;
+    }
+  }
+
+  @Override protected void getAsync(Client client, String pathIncludingQuery) {
+    client.target(url(pathIncludingQuery))
+      .request(MediaType.TEXT_PLAIN_TYPE)
+      .async()
+      .get(new InvocationCallback<String>() {
+        @Override public void completed(String response) {
+        }
+
+        @Override public void failed(Throwable throwable) {
+          Span span = tracer().currentSpan();
+          if (null != span) {
+            span.error(throwable).finish();
+          }
+          throwable.printStackTrace();
+        }
+      });
+  }
+
+  @Override
+  protected void post(Client client, String pathIncludingQuery, String body) {
+      try {
+        client.target(url(pathIncludingQuery))
+          .request(MediaType.TEXT_PLAIN_TYPE)
+          .post(Entity.text(body), String.class);
+      } catch (ProcessingException ex) {
+        Span span = tracer().currentSpan();
+        if (null != span) {
+          span.error(ex).finish();
+        }
+        throw ex;
+      }
+  }
+
+  @Override @Ignore("blind to the implementation of redirects")
+  public void redirect() {
+  }
+
+  @Override @Ignore("doesn't know the remote address")
+  public void reportsServerAddress() {
+  }
+}

--- a/instrumentation/jaxrs2/src/test/java/brave/jaxrs2/ITTracingJaxRSClientBuilder.java
+++ b/instrumentation/jaxrs2/src/test/java/brave/jaxrs2/ITTracingJaxRSClientBuilder.java
@@ -13,12 +13,10 @@
  */
 package brave.jaxrs2;
 
-import brave.Span;
 import brave.test.http.ITHttpAsyncClient;
 import org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder;
 import org.junit.Ignore;
 
-import javax.ws.rs.ProcessingException;
 import javax.ws.rs.client.Client;
 import javax.ws.rs.client.Entity;
 import javax.ws.rs.client.InvocationCallback;
@@ -32,8 +30,7 @@ public class ITTracingJaxRSClientBuilder extends ITHttpAsyncClient<Client> {
       .register(TracingClientFilter.create(httpTracing))
       .connectTimeout(1, TimeUnit.SECONDS)
       .readTimeout(1, TimeUnit.SECONDS)
-      .executorService(httpTracing.tracing().currentTraceContext()
-        .executorService(Executors.newCachedThreadPool()))
+      .executorService(currentTraceContext.executorService(Executors.newCachedThreadPool()))
       .build();
   }
 
@@ -42,17 +39,9 @@ public class ITTracingJaxRSClientBuilder extends ITHttpAsyncClient<Client> {
   }
 
   @Override protected void get(Client client, String pathIncludingQuery) {
-    try {
-      client.target(url(pathIncludingQuery))
-        .request(MediaType.TEXT_PLAIN_TYPE)
-        .get(String.class);
-    } catch (ProcessingException ex) {
-      Span span = tracer().currentSpan();
-      if (null != span) {
-        span.error(ex).finish();
-      }
-      throw ex;
-    }
+    client.target(url(pathIncludingQuery))
+      .request(MediaType.TEXT_PLAIN_TYPE)
+      .get(String.class);
   }
 
   @Override protected void getAsync(Client client, String pathIncludingQuery) {
@@ -64,10 +53,6 @@ public class ITTracingJaxRSClientBuilder extends ITHttpAsyncClient<Client> {
         }
 
         @Override public void failed(Throwable throwable) {
-          Span span = tracer().currentSpan();
-          if (null != span) {
-            span.error(throwable).finish();
-          }
           throwable.printStackTrace();
         }
       });
@@ -75,17 +60,18 @@ public class ITTracingJaxRSClientBuilder extends ITHttpAsyncClient<Client> {
 
   @Override
   protected void post(Client client, String pathIncludingQuery, String body) {
-      try {
-        client.target(url(pathIncludingQuery))
-          .request(MediaType.TEXT_PLAIN_TYPE)
-          .post(Entity.text(body), String.class);
-      } catch (ProcessingException ex) {
-        Span span = tracer().currentSpan();
-        if (null != span) {
-          span.error(ex).finish();
-        }
-        throw ex;
-      }
+    client.target(url(pathIncludingQuery))
+      .request(MediaType.TEXT_PLAIN_TYPE)
+      .post(Entity.text(body), String.class);
+  }
+
+
+  @Override @Ignore("automatic error propagation is impossible")
+  public void reportsSpanOnTransportException() {
+  }
+
+  @Override @Ignore("automatic error propagation is impossible")
+  public void addsErrorTagOnTransportException() {
   }
 
   @Override @Ignore("blind to the implementation of redirects")

